linchpin/api/__init__.py

```diff
@@ -1,28 +1,77 @@
 #!/usr/bin/env python
 
 import os
-
 from linchpin.api.invoke_playbooks import invoke_linchpin
 from linchpin.api.utils import yaml2json
+from linchpin.hooks import LinchpinHooks
+from linchpin.hooks.state import State
 
 
 class LinchpinError(Exception):
 
     def __init__(self,*args,**kwargs):
         Exception.__init__(self,*args,**kwargs)
 
-class LinchpinAPI:
+
+class LinchpinAPI(object):
 
     def __init__(self, ctx):
 
         self.ctx = ctx
         base_path = '/'.join(os.path.dirname(__file__).split("/")[0:-2])
         self.lp_path = '{0}/{1}'.format(base_path, self.ctx.cfgs['lp']['pkg'])
+        self._state = None
+        self._state_observers = []
+        self.hooks = LinchpinHooks(self)
+        self.current_target_data = {}
+
+    @property
+    def state(self):
+        """ getter function for state property of the API object. """
+        return self._state
+
+    @state.setter
+    def state(self, state):
+        # call run_hooks after state is being set
+        print("State change initiated")
+        value = state.split("::")
+        state = value[0]
+        sub_state = value[-1] if len(value) > 1 else None
+        self._state = State(state,
+                            sub_state
+                            )
+        for callback in self._state_observers:
+            callback(self._state)
+
+    def bind_to_state(self, callback):
+        self._state_observers.append(callback)
+
+    def prepare_ctx_params(self):
+
+        """
+        prepares few context parameters based on the current target_data
+        that is being set. 
+        res_file and inventory file are not found
+        These parameters are based topology name.
+        """
 
+        topology = self.current_target_data["topology"]
+        name = topology.split("/")[-1].split(".")[-2]
+        inv_file = name+".inventory"
```

Comment:
DO NOT HARDCODE THESE. If you need to have a full name for the inventory file, look at linchpin.conf as the defacto source and add a config option there. In fact, this function feels like it should be in the context.py.